certbot/interfaces.py

```diff
@@ -591,3 +599,97 @@ def add_message(self, msg, priority, on_crash=True):
 
     def print_messages(self):
         """Prints messages to the user and clears the message queue."""
+
+
+# Updater interfaces
+#
+# When "certbot renew" is run, Certbot will iterate over each lineage and check
+# if the selected installer for that lineage is a subclass of each updater
+# class. If it is and the update of that type is configured to be run for that
+# lineage, the relevant update function will be called for each domain in the
+# lineage. These functions are never called for other subcommands, so if an
+# installer wants to perform an update during the run or install subcommand, it
+# should do so when :func:`IInstaller.deploy_cert` is called.
+
+class GenericUpdater(object):
+    """Interface for update types not currently specified by Certbot.
+
+    This class allows plugins to perform types of updates that Certbot hasn't
+    defined (yet).
+
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def generic_updates(self, domain, *args, **kwargs):
+        """Perform any update types defined by the installer.
+
+        If an installer is a subclass of the class containing this method, this
+        function will always be called when "certbot renew" is run. If the
+        update defined by the installer should be run conditionally, the
+        installer needs to handle checking the conditions itself.
+
+        This method is called once for each domain.
+
+        :param str domain: domain in the lineage being updated
+
+        """
+
+
+class RenewDeployer(object):
+    """Interface for update types run when a lineage is renewed
+
+    This class allows plugins to perform types of updates that need to run at
+    lineage renewal that Certbot hasn't defined (yet).
+
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def renew_deploy(self, lineage, *args, **kwargs):
+        """Perform any update types defined by the installer.
+
+        If an installer is a subclass of the class containing this method, this
+        function will always be called when "certbot renew" is run. If the
+        update defined by the installer should be run conditionally, the
+        installer needs to handle checking the conditions itself.
+
+        This method is called once for each lineage renewed
+
+        :param lineage: Certificate lineage object that is set if certificate
+            was renewed on this run.
+        :type lineage: storage.RenewableCert
+
+        """
+
+
+class ServerTLSUpdater(object):
+    """Interface for updating a server's TLS configuration.
+
+    An installer that wants to perform TLS server updates according to this
+    interface must not only be a subclass but must respect the disablement of
+    server side TLS updates if done so by the user when performing other
+    installer functions. For example, if server TLS updates are disabled and
```

Comment:
As in, must not modify it elsewhere? We may want to restate this as "only this method can be used to modify server tls configurations", because we're disabling the entire method outside of it; they don't have to check the flag themselves.

Reply:
Only if the `--disable-server-tls-updates` is defined on the command line or in the configuration file. This interface method is intended to be run on `renew` verb, so it shouldn't be expanded for use in every TLS configuration change.

If this is defined in the command line by the user, an installer implementing this shouldn't add our TLS configuration include to the `VirtualHost` / `server` block, or configure additional security enhancements.

It acts as a general "Don't worry Certbot, I got this" - handle for the TLS configuration from behalf of the user.

Reply:
Is what you're trying to say that `--disable-server-tls-updates` should also apply to `deploy_cert`? I don't think that's what we want. And even if it was, that comment should go there, not here.

If `--disable-server-tls-updates` is `True`, will this method be run at all? We should put the answer to that in this comment explicitly.

Reply:
Yes, that was the original plan, to give user a possibility to tell Certbot that they will handle the TLS configuration (and updates to it) manually. We can discuss about this functionality further. I didn't yet change the behavior.

Currently if `--disable-server-tls-updates` (now changed to `--disable-tls-configuration-updates`) is `True`, the updater interface method doesn't get run at all, but `deploy_cert` will. The implementing Installer should keep configuring the certificate paths, but for the more detailed TLS configuration should check this CLI parameter value, and act accordingly.

Reply:
I can't for the life of me find the original plan anywhere.

It seems like since we're talking about "updates" here, that's a separate thing from `deploy_cert`, which is only called initially.

Reply:
So my initial proposal was worded like this. It was in an email called "TLS Updates Proposal", let me know if you'd like me to send it again.

I think my main concerns were making Certbot useful to more people and to make it so those who disable our TLS updates for renewal (keeping in mind there are valid reasons to do so) don't suddenly have an update forced upon them through normal Certbot usage.

If a sysadmin wants to deploy a site with a TLS config they know shouldn't be modified, but they want to use a Certbot installer plugin for automated restarts on renewal, security enhancements like slowly increasing HSTS age and OCSP prefetching, etc. forcing them to take our TLS config first and revert it back seems unnecessary. This would make automating that sites deployment very hard without a deep understanding of the changes Certbot's going to make which I wouldn't expect them to have or consider stable and unchanging at this point. I also think adding/removing domains isn't an uncommon part of Certbot usage. If you have disabled TLS security updates and need to do this, do you have to manually revert pieces of your config in this case too?

In theory, there could be a separate flag for this, but it's hard for me to imagine when you'd want one without the other and it causes problems when people forget one of the flags.

Reply:
Alternatively, they might want to manually check changes that we make. So, they'd use certbot to install the cert because they'll go check it, but wouldn't want us changing the around their config automatically thereafter. But at the same time they wouldn't want to not have us deploy it in the first place, because otherwise they have to go figure out where the certs live and what the best way to set all that up is.

Also, thunderbird seems to have eaten your original email; I only have the replies to it somehow.

Reply:
With this approach they could still do that. They could not provide `--disable-server-tls-updates` and later add it with `certbot enhance` or to a configuration file. The same isn't true if it cannot be disabled outside of `certbot renew`.

My initial email is just a git patch that doesn't include much more than what's documented in this PR, but I've sent that to you again. If you're missing any of the follow up discussion and would like to read that, let me know.

Reply:
Here are my thoughts:

This entire thing is a complicated use case that only expert Certbot users could set up successfully in the first place.

Given that, we should make the mental model as consistent as possible. It is weird that something called update applies to something that is not an update.

If they want to add/remove domains to a cert, they can run `certonly`. For any use case, they could individually run any command with or without the flag and add or remove it later.

Reply:
I want to get this and the other PRs @joohoi has been building off of merged so he's not blocked. His other work should not depend on the `ServerTLSConfigurationUpdater` class so let's land a PR based on this with all of that code removed.

We should keep the `ServerTLSConfigurationUpdater` code around somewhere in a branch though so it can be reused and maybe modified because I think getting this Updater landed is important. Without it, Apache and Nginx only update the user's TLS config when a cert is actually renewed and Postfix is unable to do so at all.

@joohoi, can you do this surgery? I'd also be interested to hear your thoughts on the discussion here, but that can wait until later if you want.

Reply:
I think that while being somewhat complicated use case, it's something that the regular Certbot users would never even stumble upon of, and the installers would just do their thing in keeping the configuration secure. 

That said, even without this interface, it's still possible for installers to handle the task using the `GenericUpdater` and `RenewDeployer`, it just can't be explicitly allowed while disabling possible other tasks that the installer might be updating.

For now, I created a new PR that is forked from this one, the only difference being the removal of the `ServerTLSConfigurationUpdater` and its functionality. The PR is #5879